# Task1:


In [1]:
import pandas as pd
import numpy as np

# 1. Получение данных

In [2]:
df= pd.read_csv('umn_foursquare_datasets\checkins.dat', sep='|', header='infer', skiprows=2 
            , names = ["id", "user_id", "venue_id", "latitude", "longitude","created_at"])

# очистим пустые
df.dropna(inplace=True)

C:\Users\semen\Anaconda2\lib\site-packages\IPython\core\interactiveshell.py:2723: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
#Анализ данных видим что некоторые данные неопределенного формата преобразуем их в нужный нам
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1021966 entries, 0 to 1021965
Data columns (total 6 columns):
id            1021966 non-null object
user_id       1021966 non-null float64
venue_id      1021966 non-null float64
latitude      1021966 non-null object
longitude     1021966 non-null object
created_at    1021966 non-null object
dtypes: float64(2), object(4)
memory usage: 54.6+ MB


In [34]:
#проработаем с данными уберем пробелы
df.latitude=df.latitude.apply(lambda x: x.strip())
df.longitude=df.longitude.apply(lambda x: x.strip())
#почистим пустые снова
df.latitude.replace('', np.nan, inplace=True)
df.longitude.replace('', np.nan, inplace=True)
df.dropna(inplace=True)

In [35]:
#неопределенного формата преобразуем их в нужный нам
df.created_at=df.created_at.apply(pd.to_datetime)
df.latitude=df.latitude.apply(pd.to_numeric)
df.longitude=df.longitude.apply(pd.to_numeric)

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 396634 entries, 1 to 1021964
Data columns (total 6 columns):
id            396634 non-null object
user_id       396634 non-null float64
venue_id      396634 non-null float64
latitude      396634 non-null float64
longitude     396634 non-null float64
created_at    396634 non-null datetime64[ns]
dtypes: datetime64[ns](1), float64(4), object(1)
memory usage: 21.2+ MB


# 2. Логика данных

Теперь необходимо кластеризовать данные координаты, чтобы выявить центры скоплений туристов. Поскольку баннеры имеют сравнительно небольшую площадь действия, нам нужен алгоритм, позволяющий ограничить размер кластера и не зависящий от количества кластеров.
**Примечание:** на 396634 строках кластеризация будет работать долго. Быть очень терпеливым не возбраняется — результат от этого только улучшится. Но для того, чтобы сдать задание, понадобится сабсет из первых 100 тысяч строк. 

In [42]:
#берем первые 100 000 строк
testData = df[:100000]
#по условию нужно использовать алгоритм MeanShift
#он работает с string or number получим данные
testData

In [48]:
testData.id = testData.id.apply(lambda x: int(x))
testData.user_id = testData.user_id.apply(lambda x: int(x))
testData.venue_id = testData.venue_id.apply(lambda x: int(x))
testData.latitude = testData.latitude.apply(lambda x: int(x))
testData.longitude  = testData.longitude.apply(lambda x: int(x))


C:\Users\semen\Anaconda2\lib\site-packages\pandas\core\generic.py:2701: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [44]:
from sklearn.cluster import MeanShift

In [46]:
model = MeanShift(bandwidth=0.1)
preds=model.fit_predict(testData)


TypeError: float() argument must be a string or a number

# 100. Дополнительно

Альтернативная обработка для неизвестного формата данных

#Это длолго и ресурсоемко обрабатывать файл

fileCheckIn = open("umn_foursquare_datasets\checkins.dat", 'r')
listCheckIn = list(fileCheckIn)
fileCheckIn.close()


isHead=0
label=list()
dataItem=list()
for line in listCheckIn:
    if isHead==0:
        label=line.strip().split('|')
        isHead=1
    else:
        if not(line.startswith('---')):
            for item in line.strip().split('|'):
                dataItem.append(item)
            
df=pd.DataFrame.from_records(dataItem, columns=label)